Funcky.SourceGenerator/OrNoneFromTryPatternGenerator.cs

```diff
@@ -40,69 +37,134 @@ private static IncrementalValueProvider<ImmutableArray<MethodPartial>> GetOrNone
+               .Where(a => a.AttributeClass?.ToDisplayString() == AttributeFullName)
+               .Where(AttributeBelongsToPartialPart(classDeclarationSyntax))
+               .Select(ParseAttribute)
+               .ToImmutableArray() is { Length: >=1 } attributes
+            ? new SemanticTarget(classDeclarationSyntax, attributes)
+            : null;
+
+    private static Func<AttributeData, bool> AttributeBelongsToPartialPart(ClassDeclarationSyntax partialPart)
+        => attribute => attribute.ApplicationSyntaxReference?.GetSyntax().Ancestors().OfType<ClassDeclarationSyntax>().FirstOrDefault() == partialPart;
+
+    private static ParsedAttribute ParseAttribute(AttributeData attribute)
+        => attribute.ConstructorArguments.Length >= 2
+           && attribute.ConstructorArguments[0].Value is INamedTypeSymbol type
+           && attribute.ConstructorArguments[1].Value is string methodName
+            ? new ParsedAttribute(type, methodName)
+            : throw new InvalidOperationException("Invalid attribute: expected a named type and a method name");
+
+    private static MethodPartial ToMethodPartial(SemanticTarget semanticTarget, Compilation compilation)
     {
-        var attribute = methodDeclaration.GetAttributeByUsedName("OrNoneFromTryPattern");
-        var compilationUnit = methodDeclaration.TryGetParentSyntax<CompilationUnitSyntax>()!;
+        var methods =
+            from attribute in semanticTarget.Attributes
+            from method in attribute.Type.GetMembers().OfType<IMethodSymbol>()
+            where method.Name == attribute.MethodName
+            select GenerateOrNoneMethod(attribute.Type, method);
+
+        return new MethodPartial(
+            NamespaceName: GetNamespaceName(semanticTarget.ClassDeclarationSyntax, compilation),
+            ClassName: semanticTarget.ClassDeclarationSyntax.Identifier.ToString(),
+            Methods: methods.ToImmutableArray(),
+            SourceTree: semanticTarget.ClassDeclarationSyntax.SyntaxTree);
+    }
 
-        return GetNamespaceName(methodDeclaration, compilation) is { } namespaceName && GetClassName(methodDeclaration) is { } className
-            ? new MethodPartial(namespaceName, className, compilationUnit.Usings.ToList(), CreateMethodImplementation(methodDeclaration, GetMethodValue(compilation, methodDeclaration, attribute), GetTypeValue(compilation, methodDeclaration, attribute)))
-            : null;
+    private static MethodDeclarationSyntax GenerateOrNoneMethod(ITypeSymbol type, IMethodSymbol method)
+        => MethodDeclaration(
+            ParseTypeName($"Funcky.Monads.Option<{type.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat)}>"),
+            GetMethodName(type, method))
+            .WithModifiers(TokenList(Token(SyntaxKind.PublicKeyword), Token(SyntaxKind.StaticKeyword)))
+            .WithParameterList(ParameterList(SeparatedList(method.Parameters.Where(p => p.RefKind == RefKind.None).Select(GenerateParameter))))
```

Comment:
There are several comparisons with a RefKind in this file, would maybe you like to use the is operator instead of ==?

Reply:
[b05b317](https://github.com/polyadic/funcky/pull/634/commits/b05b31770786edd090edefacc6f9d71a91d165ed)